text/006-bft-based-ordering-service.md

```diff
@@ -0,0 +1,405 @@
+
+Except from the configuration, the library also saves some [metadata](https://github.com/SmartBFT-Go/consensus/blob/v2/smartbftprotos/messages.proto#L105-L111) with each block that is committed.
+The overhead of the metadata is constant and negligible. 
+
+Lastly, the library writes blocks received from the leader as well as signatures on commit messages into the Write Ahead Log (WAL), however the WAL can be garbage collected after each block commit, 
+so overall the storage overhead of the WAL can be kept constant and not depend on the size of the ledger. 
+
+
+## Embedding the library in a distributed application
+
+ The [SmartBFT library](https://github.com/SmartBFT-Go/consensus) presentes a minimal interface to the application that uses it:
+
+**Input**: To submit requests to be ordered, the application invokes `SubmitRequest` and passes a byte representation of the request.
+````
+SubmitRequest(req []byte) error
+````
+
+The library offloads communication to the application that embeds it, hence it is the responsibility of the application to feed the messages from other nodes, or requests forwarded from other nodes into the library instance:
+
+````
+HandleMessage(sender uint64, m *Message)
+
+HandleRequest(sender uint64, req []byte)
+````
+
+**Output**: The library outputs requests in totally ordered batches by notifying the application via a callback:
+
+````
+Deliver(proposal bft.Proposal, signature []bft.Signature) Reconfig
+````
+
+The application is free to implement the callback as it sees fit in order to realize its desired functionality. 
+When embedded inside a blockchain application, the proposal would often denote a block constructed by the leader of that round, and the signatures array would contain a quorum of signatures collected by the library from the various nodes. The Reconfig is an object that allows the application to convey to the library to reconfigure itself (for example, to expand or reduce the number of nodes).
+
+**Dependencies**: The library interacts with the external world through the application, by delegating capabilities such as cryptography, communication and storage to the application. 
+The full list of dependencies can be found [here](https://smartbft-go.github.io/godoc/pkg/github.com/SmartBFT-Go/consensus/pkg/api/index.html), and below the important ones are listed:
+
+The SmartBFT library delegates interaction with the external world (communication and storage) to the application that consumes it. It also doesn't come with its own signature scheme and also assumes its consumer
+implements it. All such abilities it delegates to the consumer application are termed "dependencies" and their full list can be found [here](https://smartbft-go.github.io/godoc/pkg/github.com/SmartBFT-Go/consensus/pkg/api/index.html).
+The dependencies can be categorized into two sets:
```

Comment:
These two paragraphs say, more or less, the same thing.  Could they be combined?